walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -10,14 +10,79 @@
 // distributed under the License is distributed on an "AS IS" BASIS,
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
-// limitations under the License.
+// limitations under the License. 
 
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.ArrayList;
+import java.lang.Math;
 
 public final class FindMeetingQuery {
+
+  private Collection<String> mAttendees;
+  private long mDuration;
+  private final List<TimeRange> mTimeRanges = new ArrayList<>();
+  private final Collection<TimeRange> mSlots = new ArrayList<>();
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    mAttendees = request.getAttendees();
+    mDuration = request.getDuration();
+    mTimeRanges.clear();
+    mSlots.clear();
+
+    // Get all TimeRange of events containing request attendees
+    for (Event e: events){
+      for (String a: mAttendees){
+        if (e.getAttendees().contains(a)){
+          TimeRange cur = e.getWhen();
+          mTimeRanges.add(cur);
+        }
+      }
+    }
+    // Sort with ascending order
+    Collections.sort(mTimeRanges, TimeRange.ORDER_BY_START);
+
+    if (TimeRange.START_OF_DAY + mDuration > TimeRange.END_OF_DAY+1) return mSlots;
+
+    if (mTimeRanges.isEmpty()) {
+      mSlots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+    } else {
+      // Find slots before first busy event
+      if (TimeRange.START_OF_DAY + mDuration <= mTimeRanges.get(0).start()){
+        mSlots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, mTimeRanges.get(0).start(), false));
+      }
+      // Find slots between busy events
+      final int N = mTimeRanges.size();
```

Comment:
try to give a variable a name if it shows on several places.
for example: rangeSize?